pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+
+		/// Max size of a data collection
+		#[pallet::constant]
+		type MaxCollectionSize: Get<u32>;
+
+		/// Max number of collections
+		#[pallet::constant]
+		type MaxFeedersPerKey: Get<u32>;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type Collection<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		BoundedBTreeMap<T::OracleKey, (T::OracleValue, T::Moment), T::MaxCollectionSize>,
+		ValueQuery,
+	>;
+
+	/// Store the keys that are registed for this collection
+	/// Only keys registered in this stored can be used to create the collection
+	#[pallet::storage]
+	pub(crate) type Keys<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		Blake2_128Concat,
+		T::OracleKey,
+		KeyInfo<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AddedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		RemovedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		UpdatedFeeders {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		},
+		UpdatedCollection {
+			collection_id: T::CollectionId,
+			keys_updated: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account who trigger the action is not a collection admin.
+		IsNotAdmin,
+
+		/// The key is not in the collection.
+		KeyNotInCollection,
+
+		/// Collection size reached
+		MaxCollectionSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update the feeders associated to an specific key.
+		///
+		/// NOTE: In the future this should be go though change guard.
+		/// By now we trust in the admin to choose honest feeders without
+		/// invertor supervision
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(2)]
+		pub fn update_feeders(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				T::IsAdmin::check((who, collection_id)),
+				Error::<T>::IsNotAdmin
+			);
+
+			Self::mutate_and_remove_if_clean(&collection_id, &key, |info| {
+				info.feeders = feeders.clone();
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::UpdatedFeeders {
+				collection_id: collection_id,
+				key: key,
+				feeders,
+			});
+
+			Ok(())
+		}
+
+		/// Update the collection, doing the aggregation for each key in the
+		/// process
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(3)]
+		pub fn update_collection(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let values = Keys::<T>::iter_key_prefix(&collection_id)
+				.map(|key| Self::get(&key, &collection_id).map(|value| (key, value)))
+				.collect::<Result<Vec<_>, _>>()?;
```

Comment:
I am not sure we should do that...

Reply:
The storage tree should not be unbalanced as we are using the blake2 hashing for generating the prefix but I wonder whether this could bite us...

Reply:
And I would also generate the aggregates here. Otherwise this is just used for collecting the feeders for keys, but in theoru this should cache the actual final results.

Reply:
The call to 'Self::get()' already performs the aggregation

Reply:
Ahh. Okay! So this is caching everything?

Reply:
If we do not want to iterate though the keys, then we need to stores them under the same structure. By doing it, we would increase the PoV in `update_feeders()`, because it would need to write all keys each time a key is updated. 

But it could work because I think that call is not called frecuently and the admin could update all feeders for all keys in just one call.
If agree I go with it.

Reply:
Why do we want to update all feeders at once? From a UI perspective and to simlify this api it would be easier if we could just add and remove single once.

Reply:
I would estimate the same! SOunds good.

Reply:
@mustermeiszer, because there were several topics in this thread, to recap, regarding using `iter_key_prefix`, are you fine maintaining that call or do you prefer to remove it?

If we do not want that call, then the storage should be as follows: `StorageMap<CollectionId, Map<Key, Vec<Feeders>>>`, so a single addition/modification in any key or feeder will require to write/read the entire collection.

I'm not against the this because `update_collection()` is called more frequently than `update_feeders()`

Reply:
@wischli wdyt regarding iterating over storage keys vs. the above solution? 

inside which logic would the above happen @lemunozm ?

Reply:
- You would read the entire collection:
  -  each update_collection() action (more frequent).
- You would write the entire collection:
  -  each key/feeder is updated by the admin (less frecuent)
  -  each time a loan is created or closed, because the loan increases the ref counter inside of that storage. (less frequent)

Reply:
Taking into account than now in each `update_collection()` we're reading each value of each key individually (`O(K * V)`), I would go with the `iter_key_prefix()` solution, which only adds `O(k)`. So if we have a performance limitation, will be because the first `O(k * V)` is not by iterating this `O(k)`, which is a dimension lower. WDYT?

Reply:
I aggree, I am just worried that the iterator bites us. Usually, you should not use it, as there is no clear guarantee besides the hashing security that other storage does not introduce a lot of branching. 

I am not 100% sure, but I think the iteration will then have much more reads than we expect.

Reply:
Would need to take a look at the iterator implementaion again.

Reply:
But move ahead as is. And we see if we later change it.

Reply:
Ok, I would also need to dig into it